File: spec/javascripts/jsx/spec-support/logging.js
/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/* eslint-disable no-console */

let filename

export function logCurrentContext(currentContext, options = {}) {
  let messages = []
  if (options.message) {
    messages = Array.isArray(options.message) ? options.message : [options.message]
  }

  const logMessage = [...messages, 'Related spec context(s):']
  const entries = currentContext.stack.map((entry, index) => {
    let contextName = `${entry.type}: "${entry.description}"`
    if (index === currentContext.stack.length - 1 && entry.phase) {
      contextName = `${contextName} – ${entry.phase}`
    }
    return index === 0 ? `  ${contextName}` : `  ↳ ${contextName}`
  })

  console[options.logType || 'error'](logMessage.concat(entries).join('\n'))
  if (options.sourceStack) {
    console.log(options.sourceStack)
  }
}

export function logError(error, options = {}) {
  console.error(error)

  // TEMP
  if (error.filename) {
    console.log(error.filename)
    console.log(error.colno, error.lineno)
  }

  if (options.currentContext) {
    logCurrentContext(options.currentContext, {...options, logType: 'error'})
  } else if (options.message) {
    console.error(options.message)
  }
}

export function logHere(options = {}) {
  try {
    throw new Error('message')
  } catch (error) {
    logError(error, options)
  }
}

export function getStack(message) {
  try {
    throw new Error(message)
  } catch (error) {
    return extractStack(error.stack, false)
  }
}

export function logTrackers(trackers, optionFn) {
  const behaviorMap = {}
  trackers.forEach(tracker => {
    behaviorMap[tracker.type] = behaviorMap[tracker.type] || []
    behaviorMap[tracker.type].push(tracker)
  })

  Object.keys(behaviorMap).forEach(trackerType => {
    const typeTrackers = behaviorMap[trackerType]
    logCurrentContext(typeTrackers[0].currentContext, {
      ...optionFn(trackerType, typeTrackers),
      sourceStack: typeTrackers[0].sourceStack,
    })
  })
}

export function extractStack(fullStack, skipError = true) {
  if (fullStack) {
    let stack = []
    const stackLines = fullStack.split('\n')

    if (!(skipError && /^error/i.test(stackLines[0]))) {
      stack.push(stackLines.shift())
    }

    if (filename) {
      const filenameLineIndex = stackLines.findIndex(line => line.indexOf(filename !== -1))
      if (filenameLineIndex !== -1) {
        stackLines.splice(0, filenameLineIndex + 2)
      }
    }
    stack = stack.concat(stackLines)

    if (stack.length) {
      return stack.join('\n')
    }
  }

  return null
}

// Obtain filename to improve stacktrace culling.
try {
  throw new Error('obtain filename')
} catch (error) {
  if (error.stack) {
    filename = extractStack(error.stack)
      .replace(/(:\d+)+\)?/, '')
      .replace(/.+\//, '')
  } else {
    filename = ''
  }
}
/* eslint-enable no-console */

File: gems/canvas_cache/lib/redis_client/logging.rb
# frozen_string_literal: true

#
# Copyright (C) 2018 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.
#

class RedisClient
  module Logging
    COMPACT_LINE = "Redis (%{request_time_ms}ms) %{command} %{key} [%{host}]"
    NON_KEY_COMMANDS = %w[eval evalsha].freeze
    SET_COMMANDS = %w[set setex].freeze

    def call(request, config)
      client.last_command_at = start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      response = super
    ensure
      log_style = CanvasCache::Redis.log_style
      if log_style != "off" && Rails.logger
        command = request.first

        end_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
        message = {
          message: "redis_request",
          command:,
          # request_size is the sum of all the string parameters send with the command.
          request_size: request.sum { |c| c.to_s.size } - command.to_s.size,
          request_time_ms: ((end_time - start_time) * 1000).round(3),
          host: config.server_url,
        }

        unless NON_KEY_COMMANDS.include?(command)
          message[:key] = case command
                          when "mset"
                            request[1..].select.with_index { |_, i| i.even? }
                          when "mget", "scan"
                            request[1..]
                          else
                            request[1]
                          end
        end

        if defined?(Marginalia)
          message[:controller] = Marginalia::Comment.controller
          message[:action] = Marginalia::Comment.action
          message[:job_tag] = Marginalia::Comment.job_tag
        end

        if SET_COMMANDS.include?(command) && Thread.current[:last_cache_generate]
          # :last_cache_generate comes from the instrumentation added in
          # config/initializers/cache_store.rb in canvas.
          #
          # TODO: port the code in that initializer to something in thi gem.
          #
          # This is necessary because the Rails caching layer doesn't pass this
          # information down to the Redis client -- we could try to infer it by
          # looking for reads followed by writes to the same key, but this would be
          # error prone, especially since further cache reads can happen inside the
          # generation block.
          message[:generate_time_ms] = Thread.current[:last_cache_generate] * 1000
          Thread.current[:last_cache_generate] = nil
        end
        e = $!
        if e
          message[:error] = e.to_s
          message[:response_size] = 0
        else
          message[:response_size] = response&.size || 0
        end

        logline = format_log_message(message, log_style)
        Rails.logger.debug(logline)
      end
    end

    def format_log_message(message, log_style)
      if log_style == "json"
        JSON.generate(message.compact)
      else
        message[:key] ||= "-"
        COMPACT_LINE % message
      end
    end
  end
end
